prost-build/src/path.rs

```diff
@@ -1,61 +1,107 @@
 //! Utilities for working with Protobuf paths.
 
-use std::collections::HashMap;
 use std::iter;
 
 /// Maps a fully-qualified Protobuf path to a value using path matchers.
 #[derive(Debug, Default)]
 pub(crate) struct PathMap<T> {
-    matchers: HashMap<String, T>,
+    // insertion order might actually matter (to avoid warning about legacy-derive-helpers)
+    // see: https://doc.rust-lang.org/rustc/lints/listing/warn-by-default.html#legacy-derive-helpers
+    pub(crate) matchers: Vec<(String, T)>,
 }
 
 impl<T> PathMap<T> {
     /// Inserts a new matcher and associated value to the path map.
     pub(crate) fn insert(&mut self, matcher: String, value: T) {
-        self.matchers.insert(matcher, value);
+        self.matchers.push((matcher, value));
     }
 
-    /// Returns the value which matches the provided fully-qualified Protobuf path.
-    pub(crate) fn get(&self, fq_path: &'_ str) -> Option<&T> {
-        // First, try matching the full path.
-        iter::once(fq_path)
-            // Then, try matching path suffixes.
-            .chain(suffixes(fq_path))
-            // Then, try matching path prefixes.
-            .chain(prefixes(fq_path))
-            // Then, match the global path. This matcher must never fail, since the constructor
-            // initializes it.
-            .chain(iter::once("."))
-            .flat_map(|path| self.matchers.get(path))
-            .next()
+    pub(crate) fn get<'a>(&'a self, fq_path: &'_ str) -> Iter<'a, T> {
```

Comment:
I think you can elide this lifetime

`pub(crate) fn get(&self, fq_path: &str) -> Iter<'_, T>`

Reply:
Done